---
title: Usage examples for functions_covstats.R
author: "Kristen N. Finch"
date: 'Original: 12/2/2023'
output:
  html_document:
    df_print: paged
  pdf_document: default
---

## About

This is a Rmarkdown document that contains some examples for using functions from functions_cov_stats.R

The full dataset is available as supplement for a research article submitted in August 2018 and published in January 2019. Information about the study, study system, and premise for this dataset can be found the research article:

Finch, K.N., Jones, F.A. and Cronn, R.C., 2019. Genomic resources for the Neotropical tree genus *Cedrela* (Meliaceae) and its relatives. BMC Genomics, 20(1), p.58.

[DOI](https://doi.org/10.1186/s12864-018-5382-6) https://doi.org/10.1186/s12864-018-5382-6

[Read the open access article here](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-018-5382-6)

This document was prepared by Kristen N. Finch to supplement the article. The reader will find R code necessary to replicate our analysis with the files provided.

Questions about this document should be directed to Kristen N. Finch using the contact email provided under this ORCID ID: https://orcid.org/0000-0003-2098-7546. This document is not intended as a tutorial. This document is intended for advanced users. For a basic intro to R, email Kristen N. Finch. All users should frequently examine generated data frames with head(). The View() function can also be used, but viewing large datasets can cause R to crash. 

### Glossary of Terms

Term                        |Description                                                                                                                            |
-------------------------------------------------|----------------------------------------------------------------------------------------------|
Individual|an individual tree that was included in this analysis. Synonymous with specimen. 
Specimen|an individual tree that was included in this analysis. Synonymous with individual. 
Read|a sequence of nucleotides generated from Illumina sequencing. 
Gene target or target|10,001 transcripts from the transcriptome of CEOD_NYBG (C. odorata from the New York Botanical Garden) that were selected for hybridization probe design for target capture. 
Depth|synonymous with depth of coverage; this metric can be used to describe the sum of sequence reads from an idividual that mapped to the reference sequence • sequence read length ÷ count bases of reference sequence that are represented in a mapped read. 
Yield|in this case yield refers to total sequence reads generated. 
On-target yield|on-target yield specifies sequence reads that were generated and show identity to the gene targets; thus, on-target indicates mapping to the target gene sequences. 
SNP|single nucleotide polymorphism
FST|Weir-Cockerham's FST or a measurement of population differentiation. Here we calculated FST on a per marker (SNP) basis via VCFTools treating *Cedrela* species as populations. 

### Species Code Glossary

Code|Species|
----------------|----------------|
CEAN|*Cedrela angustifolia*|
CEFI|*Cedrela fissilis*|
CEMO|*Cedrela montana*|
CEOD|*Cedrela odorata*|
CESA|*Cedrela saltensis*|
GUGU|*Guarea guidonia*|
SWMA|*Swietenia mahagoni*|
TRTU|*Trichilia tuberculata*|

### Set Up

This document was prepared with R Version 4.3.2 (2023-10-31) -- "Eye Holes" RStudio version 2023.09.1+494, and the most up to date packages available. Analyses may not work as describe with different versions of R and packages. Also note that any changes to directory names or filenames will require changes to the code below. 

### Libaries

Dependencies may need to be installed manually. R will prompt. 

```{r, message=FALSE, warning=FALSE}
library(dplyr) 
library(ggplot2) 
library(RColorBrewer) 
library(reshape) #R package version 0.8.7
library(stringr) #R package version 1.3.1
```

### Coverage Statistics
Most of our analyses were based on coverage statistics generated by bbmap.sh (BBTools. DOE Joint Genome Institute. https://jgi.doe.gov/data-and-tools/bbtools/.) with parameter "covstats=<output.file>" 

Coverage Statistics output variables:

Column name                                      |Description                                                                                   |
-------------------------------------------------|----------------------------------------------------------------------------------------------|
ID|reference sequence identifier
Avg_fold|Log(mean coverage) for the reference sequence
Length|length of the reference sequence
Ref_GC|G and C content of the reference sequence
Covered_percent|percentage of the length of the reference sequence that is represented among mapped reads
Covered_bases|number of nucleotide bases of the reference sequence that is represented among mapped reads
Plus_reads|number of reads that mapped to the plus strand
Minus_reads|number of reads that mapped to the minus strand
Read_GC|G and C content of the mapped reads
Median_fold|Log(median coverage)
Std_Dev|standard deviation of coverage

#### Pre-processing Coverage Sattistics
bbmap.sh will output the coverage statistics file with a "#" as the first character or "#ID"
R cannot process a file with # because it reads # as comment. So the # must be removed from all coverage statistics files. All files provided here have been processed in the following manner in Unix:

```{r,eval=FALSE}
sed -i -e 's/#//g' *_covstats.txt 
```

### Functions
This document contains functions that were developed specifically to process coverage statistics files from bbmap.sh. 

Call|Function|Input(s)|
----------------|----------------|----------------|
get_cov_percent|Generates a dataframe with covered percent. One column per specimen.|List of data frames|
get_depth|Generates a dataframe with depth of coverage. One column per specimen.|List of data frames|
get_files|Generates a file list.|File pattern|
get_reads_mapped|Generates a dataframe with sum of mapped reads. One column per specimen.|List of data frames|
get_results|Calculates mean, min, max, and 95% confidence interval for a column of data|Note on data & column of data frame|
get_spec_ids|Uses a regular expression to generate the specimen id list from the filenames.|File list and "header"|
get_spp|Uses a regular expression to generate the species list from the filenames.|File list and "header"|
get_target_ids|Saves column 1 of coverage stats life and designates it as target ids.|File list and "header"|
make_df|Generates data frame with specimens as columns and targets as rows.|Coverage stat data & specimen ids & target ids|
make_row|Generates a row of data from a assigned value.|Note about the value & an assigned value|
make_tdf|Generates data frame with targets as columns and specimens as rows.|Coverage stat data & specimen ids & target ids & species list|
readin_files|Generated a list of data frames from a file list|File list|

#### get_files()

Generate file list.Example: get_files("*file_ending.txt")

```{r}
get_files<-function(x){
  list.files(pattern = x, full.names = TRUE)
}
```

#### readin_files()

Generate list of data frames by applying read.table() to all. 

```{r}
readin_files<-function(file_list){
  lapply(file_list,read.table,header=1)
}
```

#### get_reads_mapped()

Generate data frame where each cell is the sum of mapped reads from an individual to a reference sequence. 

For example, each cell could be the number of reads from an individual that mapped to a single gene target. 

```{r}
get_reads_mapped<-function(df_list){
  output<-list()
  for (i in 1:length(df_list)){
    output[[i]]<-mutate(df_list[[i]], total = Plus_reads + Minus_reads)
  }
  reads_mapped<-list()
  for (i in 1:length(output)){
    reads_mapped[[i]]<-cbind.data.frame(output[[i]]$total)
  }
  covstat_data<-data.frame(reads_mapped)
  covstat_data[covstat_data == "NaN"] <- 0
  return(covstat_data)
}
```

#### get_depth()

Generate data frame where each cell is:
sum of mapped reads • read length ÷ count bases of reference covered,
from an individual to a reference sequence. 

For example, each cell could be depth of coverage a single gene target for an individual. 

```{r}
get_depth<-function(df_list){
  output<-list()
  for (i in 1:length(df_list)){
    output[[i]]<-mutate(df_list[[i]], depth = ((Plus_reads + Minus_reads)*101)/Covered_bases)
  }
  read_depth<-list()
  for (i in 1:length(output)){
    read_depth[[i]]<-cbind.data.frame(output[[i]]$depth)
  }
  covstat_data<-data.frame(read_depth)
  covstat_data[covstat_data == "NaN"] <- 0
  return(covstat_data)
}
```

#### get_cov_percent()

Generate data frame where each cell is the percent of the reference that is covered by mapped reads from an individual.

For example, each cell could be the percent that a single gene target is covered by mapped reads from an individual. 

```{r}
get_cov_percent<-function(df_list){
  cp_cov_per<-list()
  for (i in 1:length(df_list)){
    cp_cov_per[[i]]<-cbind.data.frame(df_list[[i]]$Covered_percent)
  }
  covstat_data<-data.frame(cp_cov_per)
  return(covstat_data)
}
```

#### get_spec_ids()

Extracts a list of specimen ids from the filenames with a regular expression. 

Header must follow character syntax with double quotes: "header"

```{r}
get_spec_ids<-function(file_list,header){
  spec_ids<-data.frame(str_extract(file_list,"[C,G,S,T]...[_]\\d{0,3}"))
  names(spec_ids)<-header
  return(spec_ids)
}
```

#### get_spp()

Extracts a list of species from the filenames with a regular expression. 

Header must follow character syntax with double quotes: "header"

```{r}
get_spp<-function(file_list,header){
  spp<-data.frame(str_extract(file_list,"[C,G,S,T]..."))
  names(spp)<-header
  return(spp)
}
```

#### get_target_ids()

Generates a list of target ids from the first column of the first file in the file list. 

Header must follow character syntax with double quotes: "header"

```{r}
get_target_ids<-function(file_list,header){
  file_1<-read.table(file_list[1],header = 1)
  target_ids<-data.frame(file_1[1])
  names(target_ids)<-header
  return(target_ids)
}
```

#### make_df()

Generates data frame with specimens as columns and targets as rows.

Example:

target_id|CEFI_9|CESA_75|
----------|------|-------|
ceod000050|    49|   2521|
ceod000075|    38|   1347|
ceod000084|    71|   3106|
ceod000163|   103|   5848|
ceod000194|    56|   2130|
ceod000233|    69|   4826|

```{r}
make_df<-function(covstat_data,spec_ids,target_ids){
  spec_ids<-t(spec_ids)
  colnames(covstat_data)<-spec_ids
  df<-cbind.data.frame(target_ids,covstat_data)
  return(df)
}
```

#### make_tdf()

Generates data frame with targets as columns and specimens as rows.

Example:

specimen|species|ceod000050|ceod000075|
|----------|----------|----------|----------|
CEFI_9  | CEFI  |      49  |      38|
CESA_75 |  CESA |     2521 |     1347|
CEFI_112|   CEFI|        61|        29|
CEFI_140|   CEFI|      1173|       708|
CEOD_202|   CEOD|       540|       217|
CEFI_230|   CEFI|       339|       201|

```{r}
make_tdf<-function(covstat_data,spec_ids,target_ids,spp){
  tcovstat_data<-t(covstat_data)
  target_ids<-t(target_ids)
  colnames(tcovstat_data)<-target_ids
  tdf<-cbind.data.frame(spec_ids,spp,tcovstat_data)
  return(tdf)
}
```

#### get_results()

Calculates mean, min, max, and 95% confidence interval for a column of data|Note on data & column of data frame.

Example Usage:
ind_percent_ontarget<-get_results("individual percent on target",yield_tdf_full$ind_percent_ontarget)
Note_about_stat must follow character syntax with double quotes: "Note about stat"

Example Output:

note_about_stat|mean  |min       |max       |err       |up        |lo        |
|----------|----------|----------|----------|----------|----------|----------|
individual percent on target| 19.75343 |0.8108261 |41.42799 |5.41383 |25.16726 |14.3396|


```{r}
get_results<-function(note_about_stat,column_of_data){
  mean<-mean(column_of_data,na.rm = TRUE)
  min<-min(column_of_data,na.rm = TRUE)
  max<-max(column_of_data,na.rm = TRUE)
  err<-qt(0.975,df=length(column_of_data)-1)*sd(column_of_data,na.rm = TRUE)/sqrt(length(column_of_data))
  up<-mean+err
  lo<-mean-err
  note_about_stat<-note_about_stat
  results<-cbind.data.frame(note_about_stat,mean,min,max,err,up,lo)
  return(results)
}
```

#### make_row()

Generates a row of data from a assigned value.

Example Usage: 
make_row("detected variants",detected_variants)
Note must follow character syntax with double quotes: "Note about value"

Example Output:

result           |stat  |
-----------------|------|
detected variants|444979|

```{r}
make_row<-function(note,stat){
  row<-cbind.data.frame(note, stat)
  names(row)<-c("result","stat")
  return(row)
}
```

### Target Capture Efficiency/Bias Across *Cedrela* and Other Meliaceae

For this analysis, we mapped the 1 million sequence reads from each individual to the gene targets to assess target capture efficiency and potential ascertainment bias. 

```{r, warning=FALSE,message=FALSE}
# the coverage stats files for this part of the analysis are in a different directory.
setwd("/Users/kristenfinch/Library/Mobile Documents/com~apple~CloudDocs/Finch_etal_2018_supplement/R_analysis/covstats_1Msubset_gene_targets/")
#IT IS NECESSARY TO CHANGE THE PATH HERE TO WHERE THE FILE IS LOCATED ON YOUR COMPUTER

#see descriptions of functions to undertand the following lines of code. 
filenames<-get_files("*_covstats.txt")

ldf<-readin_files(filenames)

read_depth<-get_depth(ldf)

spec_ids<-get_spec_ids(filenames,"specimen")

spp<-get_spp(filenames,"species")

target_ids<-get_target_ids(filenames,"target_id")

eff_df<-make_df(read_depth,spec_ids,target_ids)

eff_tdf<-make_tdf(read_depth,spec_ids,target_ids,spp)

#depth/individual
ind_mean_depth<-data.frame(eff_tdf[c(1:2)])
ind_mean_depth$ind_min_depth<-apply(eff_tdf[c(3:10003)],1,min)
ind_mean_depth$ind_max_depth<-apply(eff_tdf[c(3:10003)],1,max)
ind_mean_depth$ind_mean_depth<-apply(eff_tdf[c(3:10003)],1,mean)

#calculate individual mean depth
(mean_ind_mean_depth<-get_results("average mean depth per individual",ind_mean_depth$ind_mean_depth))

#calculate mean depth for each species
spp_mean_depth<-ind_mean_depth[c(2:5)]%>%group_by(species)%>%summarise_all(funs(mean))
names(spp_mean_depth)<-c("species","species_min_depth","species_max_depth","species_mean_depth")
(spp_mean_depth)

#depth/target
#calculate mean depth of each target
target_mean_depth<-data.frame(eff_df[1])
target_mean_depth$target_min_depth<-apply(eff_df[c(2:43)],1,min)
target_mean_depth$target_max_depth<-apply(eff_df[c(2:43)],1,max)
target_mean_depth$target_mean_depth<-apply(eff_df[c(2:43)],1,mean)

(mean_target_mean_depth<-get_results("average mean depth per target",target_mean_depth$target_mean_depth))

#transform the data from short form to long form "quickly" for plotting. Should take nearly 2 minutes.
depths<-melt(eff_tdf)
#calculate log2 for depth to facilitate plotting
depths$lval<-log2(depths$value)

```

#### Target Capture Efficiency/Bias *t* Test Cedrela v. Other Meliaceae

```{r}
ind_mean_depth$genus<-str_extract(ind_mean_depth$species,"[C,G,S,T]")
mean_depth_genus<-cbind.data.frame(ind_mean_depth$genus,ind_mean_depth$ind_mean_depth)
names(mean_depth_genus)<-c("genus","mean_depth")

test_ced<-filter(mean_depth_genus,genus == "C")
test_other<-filter(mean_depth_genus,genus != "C")
test_ced<-test_ced[2]
test_other<-test_other[2]

t.test(test_ced, test_other, alternative = "two.sided")

test_ced$genus<-"C"
test_other$genus<-"O"
test<-rbind.data.frame(test_ced,test_other)
ggplot(aes(x=genus,y=mean_depth),data=test)+
  scale_x_discrete(labels=c("C"="Cedrela","O"="Other Meliaceae"))+
  geom_boxplot()+
  theme_classic()+
  xlab("Genus")+ylab("Reads On-Target")
```

### Target Capture Efficiency/Bias by Species

#### Fig. 3.
Density plot showing depth of coverage (log2 scale) distribution across enriched targets for each species. Species distributions are color coded by species.

```{r, warning=FALSE,message=FALSE}

#color palette for colorblindness accessibility
cbPalette <- c("#009E73","#D55E00","#56B4E9","#CC79A7","#0072B2","#999999", "#E69F00","#000000")

ggplot(aes(x=lval,color=species),data=depths)+
  geom_density(size=1)+
  scale_color_manual(values=cbPalette,
                     labels=c("CEAN"="C. angustifolia","CEFI"="C. fissilis",
                              "CEMO"="C. montana","CEOD"="C. odorata",
                              "CESA"="C. saltensis","GUGU"="G. guidonia",
                              "SWMA"="S. mahagoni","TRTU"="T. tuberculata"),
                     name="Species")+
  theme_classic()+theme(plot.background = element_blank(),
                        legend.text = element_text(face = "italic",size=6.5),
                        legend.key.size = unit(3, "mm"),
                        legend.position = c(0.75,0.75),
                        legend.background = element_blank(),
                        legend.title = element_text(size=8),
                        axis.text.x=element_text(size=6.5),
                        axis.title.x=element_text(size=8),
                        axis.text.y=element_text(size=6.5),
                        axis.title.y=element_text(size=8))+
  labs(x=expression(Log[2]*" Depth of Coverage"),y="Kernal Density Estimate")
```

## Update Log

Original submission: 12/2/2023. Submitted as public resource on R_misc github repo.

Updated: 

NA

```{r,message=FALSE,warning=FALSE}
sessionInfo()
```