#Functions for processing coverage statistics generated by bbmap.sh (BBTools.
#DOE Joint Genome Institute. https://jgi.doe.gov/data-and-tools/bbtools/.) with
#parameter "covstats=<output.file>"

#These functions were originally published as supplement for: Finch, K.N.,
#Jones, F.A. and Cronn, R.C., 2019. Genomic resources for the Neotropical tree
#genus *Cedrela* (Meliaceae) and its relatives. BMC Genomics, 20(1), p.58. [Read
#the open access article
#here](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-018-5382-6)

#And was published as a full dataset:Kristen N. Finch. 2018. Dataset for genomic
#resources for the neotropical tree genus Cedrela (Meliaceae) and its relatives
#[Data set]. Oregon State University, Corvallis, OR, USA.
#https://doi.org/10.7267/NV935820Q

###Glossary of Terms

#Individual - an individual tree that was included in this analysis. Synonymous
#with specimen.

#Specimen - an individual tree that was included in this analysis. Synonymous
#with individual.

#Read - a sequence of nucleotides generated from Illumina sequencing. 

#Gene target or target - 10,001 transcripts from the transcriptome of CEOD_NYBG (C.
#odorata from the New York Botanical Garden) that were selected for
#hybridization probe design for target capture.

#Depth - synonymous with depth of coverage; this metric can be used to describe
#the sum of sequence reads from an idividual that mapped to the reference
#sequence • sequence read length ÷ count bases of reference sequence that are
#represented in a mapped read.

#Yield - in this case yield refers to total sequence reads generated. 

#On-target yield - on-target yield specifies sequence reads that were generated
#and show identity to the gene targets; thus, on-target indicates mapping to the
#target gene sequences.

#SNP - single nucleotide polymorphism

#FST - Weir-Cockerham's FST or a measurement of population differentiation. Here
#we calculated FST on a per marker (SNP) basis via VCFTools treating *Cedrela*
#species as populations.

###Species Code Glossary

#Code:Species:
#CEAN:Cedrela angustifolia:
#CEFI:Cedrela fissilis:
#CEMO:Cedrela montana:
#CEOD:Cedrela odorata:
#CESA:Cedrela saltensis:
#GUGU:Guarea guidonia:
#SWMA:Swietenia mahagoni:
#TRTU:Trichilia tuberculata:

#Dependencies may need to be installed manually. R will prompt. 

library(dplyr) 
library(ggplot2) 
library(reshape) 
library(stringr) 

###Coverage Statistics
#Most of our analyses were based on coverage statistics generated by bbmap.sh
#(BBTools. DOE Joint Genome Institute.
#https://jgi.doe.gov/data-and-tools/bbtools/.) with parameter
#"covstats=<output.file>"

#Coverage Statistics output variables:
  
#Column name|Description                                                                                   |
#ID|reference sequence identifier
#Avg_fold|Log(mean coverage) for the reference sequence
#Length|length of the reference sequence
#Ref_GC|G and C content of the reference sequence
#Covered_percent|percentage of the length of the reference sequence that is represented among mapped reads
#Covered_bases|number of nucleotide bases of the reference sequence that is represented among mapped reads
#Plus_reads|number of reads that mapped to the plus strand
#Minus_reads|number of reads that mapped to the minus strand
#Read_GC|G and C content of the mapped reads
#Median_fold|Log(median coverage)
#Std_Dev|standard deviation of coverage
####Pre-processing Coverage Sattistics

#bbmap.sh will output the coverage statistics file with a "#" as the first
#character or "#ID" R cannot process a file with # because it reads # as
#comment. So the # must be removed from all coverage statistics files. All files
#provided here have been processed in the following manner in Unix:

#sed -i -e 's/#//g' *_covstats.txt 

###Functions----

#This document contains functions that were developed specifically to process
#coverage statistics files from bbmap.sh (BBTools. DOE Joint Genome Institute.
#https://jgi.doe.gov/data-and-tools/bbtools/.)

#get_cov_percent:Generates a dataframe with covered percent. One column per
#specimen.:List of data frames:

#get_depth:Generates a dataframe with depth of coverage. One column per
#specimen.:List of data frames:

#get_files:Generates a file list.:File pattern:

#get_reads_mapped:Generates a dataframe with sum of mapped reads. One column per
#specimen.:List of data frames:

#get_results:Calculates mean, min, max, and 95% confidence interval for a column
#of data:Note on data & column of data frame:

#get_spec_ids:Uses a regular expression to generate the specimen id list from
#the filenames.:File list and "header":

#get_spp:Uses a regular expression to generate the species list from the
#filenames.:File list and "header":

#get_target_ids:Saves column 1 of coverage stats life and designates it as
#target ids.:File list and "header":

#make_df:Generates data frame with specimens as columns and targets as
#rows.:Coverage stat data & specimen ids & target ids:

#make_row:Generates a row of data from a assigned value.:Note about the value &
#an assigned value:

#make_tdf:Generates data frame with targets as columns and specimens as
#rows.:Coverage stat data & specimen ids & target ids & species list:

#readin_files:Generated a list of data frames from a file list:File list:

####get_files()

#Generate file list.Example: get_files("*file_ending.txt")

get_files<-function(x){
list.files(pattern = x, full.names = TRUE)
}

####readin_files()

#Generate list of data frames by applying read.table() to all. 

readin_files<-function(file_list){
lapply(file_list,read.table,header=1)
}


####get_reads_mapped()

#Generate data frame where each cell is the sum of mapped reads from an
#individual to a reference sequence.

#For example, each cell could be the number of reads from an individual that
#mapped to a single gene target.

get_reads_mapped<-function(df_list){
output<-list()
for (i in 1:length(df_list)){
output[[i]]<-mutate(df_list[[i]], total = Plus_reads + Minus_reads)
}
reads_mapped<-list()
for (i in 1:length(output)){
reads_mapped[[i]]<-cbind.data.frame(output[[i]]$total)
}
covstat_data<-data.frame(reads_mapped)
covstat_data[covstat_data == "NaN"] <- 0
return(covstat_data)
}

####get_depth()

#Generate data frame where each cell is:
#sum of mapped reads • read length ÷ count bases of reference covered,
#from an individual to a reference sequence. 

#For example, each cell could be depth of coverage a single gene target for an individual. 

get_depth<-function(df_list){
output<-list()
for (i in 1:length(df_list)){
output[[i]]<-mutate(df_list[[i]], depth = ((Plus_reads + Minus_reads)*101)/Covered_bases)
}
read_depth<-list()
for (i in 1:length(output)){
read_depth[[i]]<-cbind.data.frame(output[[i]]$depth)
}
covstat_data<-data.frame(read_depth)
covstat_data[covstat_data == "NaN"] <- 0
return(covstat_data)
}

####get_cov_percent()

#Generate data frame where each cell is the percent of the reference that is
#covered by mapped reads from an individual.

#For example, each cell could be the percent that a single gene target is
#covered by mapped reads from an individual.

get_cov_percent<-function(df_list){
cp_cov_per<-list()
for (i in 1:length(df_list)){
cp_cov_per[[i]]<-cbind.data.frame(df_list[[i]]$Covered_percent)
}
covstat_data<-data.frame(cp_cov_per)
return(covstat_data)
}

####get_spec_ids()

#Extracts a list of specimen ids from the filenames with a regular expression. 

#Header must follow character syntax with double quotes: "header"

get_spec_ids<-function(file_list,header){
spec_ids<-data.frame(str_extract(file_list,"[C,G,S,T]...[_]\\d{0,3}"))
names(spec_ids)<-header
return(spec_ids)
}

####get_spp()

#Extracts a list of species from the filenames with a regular expression. 

#Header must follow character syntax with double quotes: "header"

get_spp<-function(file_list,header){
spp<-data.frame(str_extract(file_list,"[C,G,S,T]..."))
names(spp)<-header
return(spp)
}

####get_target_ids()

#Generates a list of target ids from the first column of the first file in the file list. 

#Header must follow character syntax with double quotes: "header"

get_target_ids<-function(file_list,header){
file_1<-read.table(file_list[1],header = 1)
target_ids<-data.frame(file_1[1])
names(target_ids)<-header
return(target_ids)
}

####make_df()

#Generates data frame with specimens as columns and targets as rows.

#Example:

#target_id:CEFI_9:CESA_75:
#----------:------:-------:
#ceod000050:    49:   2521:
#ceod000075:    38:   1347:
#ceod000084:    71:   3106:
#ceod000163:   103:   5848:
#ceod000194:    56:   2130:
#ceod000233:    69:   4826:

make_df<-function(covstat_data,spec_ids,target_ids){
spec_ids<-t(spec_ids)
colnames(covstat_data)<-spec_ids
df<-cbind.data.frame(target_ids,covstat_data)
return(df)
}

####make_tdf()

#Generates data frame with targets as columns and specimens as rows.

#Example:

#specimen:species:ceod000050:ceod000075:
#:----------:----------:----------:----------:
#CEFI_9  : CEFI  :      49  :      38:
#CESA_75 :  CESA :     2521 :     1347:
#CEFI_112:   CEFI:        61:        29:
#CEFI_140:   CEFI:      1173:       708:
#CEOD_202:   CEOD:       540:       217:
#CEFI_230:   CEFI:       339:       201:

make_tdf<-function(covstat_data,spec_ids,target_ids,spp){
tcovstat_data<-t(covstat_data)
target_ids<-t(target_ids)
colnames(tcovstat_data)<-target_ids
tdf<-cbind.data.frame(spec_ids,spp,tcovstat_data)
return(tdf)
}

#Usage
#Usage Examples Found: 
#https://doi.org/10.7267/NV935820Q

